lib/cea/cea_decoder.js

```diff
@@ -0,0 +1,293 @@
+    // algorithm isn't stable. This sort breaks ties based on receive order.
+    this.ccPacketArray_.sort(
+        /**
+       * Stable sorting function.
+       * @param {!shaka.cea.Cea608DataChannel.Cea608Packet} ccPacket1
+       * @param {!shaka.cea.Cea608DataChannel.Cea608Packet} ccPacket2
+       * @return {!number}
+       */
+        (ccPacket1, ccPacket2) => {
+          const diff = ccPacket1.pts - ccPacket2.pts;
+          const isEqual = diff === 0;
+          return isEqual ? ccPacket1.order - ccPacket2.order : diff;
+        });
+
+    for (const ccPacket of this.ccPacketArray_) {
+      // Only consider packets that are NTSC line 21 (CEA-608).
+      // Types 2 and 3 contain DVTCC data, for a future CEA-708 decoder.
+      if (ccPacket.type === shaka.cea.CeaDecoder.NTSC_CC_FIELD_1 ||
+          ccPacket.type === shaka.cea.CeaDecoder.NTSC_CC_FIELD_2) {
+        const parsedClosedCaption = this.decodeCea608_(ccPacket);
+        if (parsedClosedCaption) {
+          parsedClosedCaptions.push(parsedClosedCaption);
+        }
+      }
+    }
+
+    this.clearExtractedPackets_();
+    return parsedClosedCaptions;
+  }
+
+  /**
+   * Decodes a CEA-608 closed caption packet based on ANSI/CEA-608.
+   * @param {shaka.cea.Cea608DataChannel.Cea608Packet} ccPacket
+   * @return {?shaka.cea.ICaptionDecoder.ClosedCaption}
+   * @private
+   */
+  decodeCea608_(ccPacket) {
+    const fieldNum = ccPacket.type;
+    let b1 = ccPacket.ccData1;
+    let b2 = ccPacket.ccData2;
+
+    // If this packet is a control code, then it also sets the channel.
+    // For control codes, cc_data_1 has the form |P|0|0|1|C|X|X|X|.
+    // "C" is the channel bit. It indicates whether to set C2 active.
+    if (shaka.cea.Cea608DataChannel.isControlCode(b1)) {
+      const channelNum = (b1 >> 3) & 0x01; // Get channel bit.
+
+      // Change the stream based on the field, and the new channel
+      if (fieldNum === 0) {
+        this.currentField1Channel_ = channelNum;
+      } else {
+        this.currentField2Channel_ = channelNum;
+      }
+    }
+
+    // Get the correct stream for this caption packet (CC1, ..., CC4)
+    const selectedChannel = fieldNum ?
+        this.currentField2Channel_ : this.currentField1Channel_;
+    const selectedMode = `CC${(fieldNum << 1) | selectedChannel + 1}`;
+    const selectedStream = this.cea608ModeToStream_.get(selectedMode);
+
+    // If on field 1, check if there are too many bad frames
+    if (fieldNum === 0) {
+      if ((b1 === 0xff && b2 === 0xff) || (!b1 && !b2)) {
+        // Completely invalid frame.  FCC says after 45 of them, reset.
+        if (++this.badFrames_ >= 45) {
+          this.reset();
+        }
+      }
+    }
+
+    // Validate the parity, cc_data_1 and cc_data_2 should be odd parity uimsbf.
+    if (!this.isOddParity_(b1) || !this.isOddParity_(b2)) {
+      return null;
+    }
+
+    // Remove the MSB (parity bit).
+    ccPacket.ccData1 = (b1 &= 0x7f);
```

Comment:
Seems weird to read b1 and b2 out of ccPacket, then modify them both again here.  At first I thought it was a typo, and that you must surely mean `ccData1 = (b1 & 0x7f)`, but I think you must need to modify both because both continue to be referenced after this point.

It might be clearer if you only modify one of the other, or even better if you simplify this method to only reference one set of data.